(new) Plugins: kernel|initrd (#5814)
canonical/snapcraft: docs/reference/plugins/kernel_plugin.rst

.. _reference-kernel-plugin:

Kernel plugin
==============

The Kernel plugin simplifies the building of kernel snaps and kernel modules.
It helps with both Ubuntu-specific kernels as well as kernels from some

> arbitrary upstream source, such as the mainline or linux-next trees.

Reviewer: ```suggestion
It helps with both Ubuntu-specific kernels as well as kernels from
other upstream sources, such as the mainline or linux-next trees.
```

The kernel build process tends to be highly repetitive and deterministic, where

> most desired modifications include applying patches and modifying Kconfig

Author: Kconfig is a [meaningful term](https://docs.kernel.org/kbuild/kconfig-language.html) in this context, but we can switch to kernel configuration or config. I was trying to draw a difference between a configuration file (a generated `.config` consumed by the Kbuild system) versus a configuration value (a `kconfig` used by the Kbuild system for generating a `.config`). But perhaps we don't need to be so hair splitting here.

> most desired modifications include applying patches and modifying Kconfig

Reviewer: To a newbie it feels like turbo encabulator jargon, but I'll defer to your judgement.

values. This plugin allows for the easy toggling of Kconfig values and entire
defconfigs used during builds to improve kernel snap iterability and

> maintainability.

Reviewer: Exactly what the reader needs to see. A great motivation to use it!

Keys
----

This plugin provides the following unique keys.

kernel-kdefconfig
~~~~~~~~~~~~~~~~~

**Type**: list of strings

**Default**: ``["defconfig"]``

The kernel configuration(s) to use when generating a ``.config``, such as those
found in ``arch/${arch}/configs`` or ``kernel/configs``.

kernel-kconfigflavour
~~~~~~~~~~~~~~~~~~~~~

**Type**: list of strings

**Default**: ``["generic"]``

The Ubuntu kernel flavor to use when generating an Ubuntu kernel config. Valid

> names are things like ``lowlatency`` and ``generic``. A list of some variants

Author: I'd say that `config` is a generally accepted reference here, but it points out the "kernel configuration file" so I will use that verbiage.

can be found `here <https://ubuntu.com/kernel/variants>`_, but valid choices
will depend on the chosen kernel source. If something other than ``generic`` is
provided, this option wins over ``kernel-kdefconfig``.

kernel-kconfigs
~~~~~~~~~~~~~~~

**Type**: list of strings

**Default**: ``[]``

An explicit list of Kconfig values to force; this will override the configs
that are set in the base configuration through ``kernel-kdefconfig`` and
``kernel-kconfigflavour``. Dependent Kconfigs will be fixed using the kbuild

> system to regenerate a valid configuration.

Reviewer: ```suggestion
A list of kernel config values to enforce. If set, this list overrides the values
in the base configurations established by the ``kernel-kdefconfig`` and
``kernel-kconfigflavour`` keys. The kernel build system is used to resolve any 
configuration dependencies.
```

Reviewer 2: This made me realize, shouldn't it be `kconfigflavor`, en-US style?

Author: @medubelko I've taken most of the suggestions and modified some phrasing slightly to express a similar intent as the original.

@bepri It *should*, but I don't know if the Kernel team plans on changing from flavour to flavor :) the option is intended to match Ubuntu kernels as flavors only make sense in that context.

Some websites are still inconsistent with it as well; see [here](https://ubuntu.com/kernel) versus [here](https://ubuntu.com/kernel/variants).

Indeed, even our most recent trees are inconsistent:

```
debian.hwe-6.14/control.d/flavour-control.stub:# This file describes the template for packages that are created for each flavour
debian.hwe-6.14/changelog:    - [Packaging] bpftool: move to separate unversioned and unflavored binary
```

I don't care either way.

kernel-enable-zfs-support
~~~~~~~~~~~~~~~~~~~~~~~~~

**Type**: bool

**Default**: ``false``

Enabling this option will build the ZFS kernel modules alongside the kernel.

kernel-enable-perf
~~~~~~~~~~~~~~~~~~

**Type**: bool

**Default**: ``false``

Enabling this option will build the ``perf`` binary.

Environment variables
---------------------

> This plugin responds to its own special variables.

Reviewer: ```suggestion
This plugin provides the following environment variables.
```

Author: Damn you, [unreliable prior art](https://documentation.ubuntu.com/snapcraft/latest/reference/plugins/uv_plugin/#environment-variables)!

This value specifies the name of the target kernel image to build, and the
value is target-architecture dependent:

* armhf: ``zImage``
* arm64, riscv64: ``Image``
* powerpc, ppc64el: ``vmlinux.strip``
* x86, x86_64, s390x: ``bzImage``

KERNEL_TARGET
~~~~~~~~~~~~~

**Default:** Depends on ``${CRAFT_ARCH_BUILD_FOR}``

This value specifies the additional targets to build, and the value is target-architecture dependent:

* x86, x86_64: ``modules``
* armhf, arm64, powerpc, ppc64el, riscv64, s390x: ``modules dtbs``

CROSS_COMPILE
~~~~~~~~~~~~~

**Default:** ``${CRAFT_ARCH_TRIPLET_BUILD_FOR}-``

This value specifies the prefix for the (cross-)compiler.

Dependencies
------------

The plugin specifies the most common build-time requirements as ``build-packages``
itself, but depending on the enabled Kconfigs some additional ones may be required.

Requirements such as ``bash`` and ``perl`` usually come from the build
environment, and other requirements may include ``pahole``.

How it works
------------

During the build step the plugin performs the following actions:

#. Pass a collection of flags built from the selected options to a kernel build

>    script within the snapcraft snap.

Reviewer: ```suggestion
   script within the Snapcraft snap.
```
